### PR TITLE
BookManagement: Use Hibernate as the persistence layer to Servlet project ( Create Model, Dao, Servlet for book, author and Category to manage Book)

### PR DESCRIPTION
**Pull request by ces-totrinh**
hi Mr.Hong, I updated my work, Please review! Thank you

### CODE REVIEW

File: src/net/bookmanagement/dao/AuthorDAO.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+import org.hibernate.Session;
+import org.hibernate.Transaction;
+
+import net.bookmanagement.model.Author;
+import net.bookmanagement.utils.HibernateUtil;
+
+public class AuthorDAO {
+
+	public void saveAuthor(Author author) {
+        Transaction transaction = null;
```
Comment (ces-hongvo) [inline src/net/bookmanagement/dao/AuthorDAO.java]: Please fix the indent

File: src/net/bookmanagement/dao/AuthorDAO.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+            if (transaction != null) {
+                transaction.rollback();
+            }
+            e.printStackTrace();
+        }
+        session.close();
+    }
+	
+	public void updateAuthor(Author author) {
+        Transaction transaction = null;
```
Comment (ces-hongvo) [inline src/net/bookmanagement/dao/AuthorDAO.java]: Please fix the indent

File: src/net/bookmanagement/dao/AuthorDAO.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+	public Author getAuthorByName(String name) {
+		Transaction transaction = null;
+		Author author = null;
+		Session session = HibernateUtil.getSessionFactory().openSession();
+		try{
+			transaction = session.beginTransaction();
+			author = (Author) session.createQuery("from Author where author_name=:name").setParameter("name", name).uniqueResult();
+			 
+			//commit transaction
+			transaction.commit();
```
Comment (ces-hongvo) [inline src/net/bookmanagement/dao/AuthorDAO.java:93]: This comment is unnecessary. Please consider removing it

File: src/net/bookmanagement/dao/AuthorDAO.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+				 transaction.rollback();
+			 }
+			 e.printStackTrace();
+		 }
+		 session.close();
+		 return listOfAuthor;
+	}
+	
+}
+
```
Comment (ces-hongvo) [inline src/net/bookmanagement/dao/AuthorDAO.java:127]: Be aware of the return null for the `List`. When we use some method like size(), might cause NPE

File: src/net/bookmanagement/dao/BookDAO.java [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+import org.hibernate.Session;
+import org.hibernate.Transaction;
+
+import net.bookmanagement.model.Book;
+import net.bookmanagement.utils.HibernateUtil;
+
+public class BookDAO {
+	public void saveBook(Book book) {
+        Transaction transaction = null;
```
Comment (ces-hongvo) [inline src/net/bookmanagement/dao/BookDAO.java]: Fix the indent

File: src/net/bookmanagement/dao/BookDAO.java [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+				 transaction.rollback();
+			 }
+			 e.printStackTrace();
+		 }
+		 session.close();
+		 return listOfBook;
+	}
+	
+}
+
```
Comment (ces-hongvo) [inline src/net/bookmanagement/dao/BookDAO.java:139]: BE aware of returning null for the List

File: src/net/bookmanagement/dao/CategoryDAO.java [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+import org.hibernate.Session;
+import org.hibernate.Transaction;
+
+import net.bookmanagement.model.Category;
+import net.bookmanagement.utils.HibernateUtil;
+
+public class CategoryDAO {
+	
+	public void saveCategory(Category category) {
+        Transaction transaction = null;
```
Comment (ces-hongvo) [inline src/net/bookmanagement/dao/CategoryDAO.java]: Fix the indent

File: src/net/bookmanagement/model/Author.java [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+	public Set<Book> get_books() {
+		return _books;
+	}
+
+	public void set_books(Set<Book> _books) {
+		this._books = _books;
+	}
+	
+}
```
Comment (ces-hongvo) [inline src/net/bookmanagement/model/Author.java:61]: Please update the order of the modifications: Public Protected Private

File: src/net/bookmanagement/model/Book.java [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+	@Id
+	@GeneratedValue(strategy = GenerationType.IDENTITY)
+	@Column(name="book_id")
+	private int _bookId;
+	
+	@Column(name="book_name")
+	private String _bookName;
+
+	@ManyToMany(fetch = FetchType.EAGER, cascade = CascadeType.ALL)
+    @JoinTable(name = "author_book_tbl",
```
Comment (ces-hongvo) [inline src/net/bookmanagement/model/Book.java]: Please fix the indent

File: src/net/bookmanagement/model/Book.java [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+	}
+	
+    public String getAuthorNames(Set<Author> authors) {		
+		String names = "";
+        for (Author author : authors){ 
+            names += author.get_authName() + ", ";
+        }
+		return names.substring(0, names.length()-2);     
+    }
+}
```
Comment (ces-hongvo) [inline src/net/bookmanagement/model/Book.java:94]: Update the order of the modifications

File: src/net/bookmanagement/model/Category.java [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+	@GeneratedValue(strategy = GenerationType.IDENTITY)
+	@Column(name="category_id")
+	private int _categId;
+	
+	@Column(name="category_name")
+	private String _categName;
+	
+	@OneToMany(mappedBy = "_category", cascade = CascadeType.ALL)
+    private Set<Book> _books;
+
```
Comment (ces-hongvo) [inline src/net/bookmanagement/model/Category.java]: Fix the indent

File: src/net/bookmanagement/servlet/BookServlet.java [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+	private AuthorDAO authorDao;
+
+    public void init() {
+    	bookDao = new BookDAO();
+    	categoryDao = new CategoryDAO();
+    	authorDao = new AuthorDAO();
+    }
+
+    protected void doPost(HttpServletRequest request, HttpServletResponse response)
+    throws ServletException, IOException {
```
Comment (ces-hongvo) [inline src/net/bookmanagement/servlet/BookServlet.java]: Fix the indent

File: src/net/bookmanagement/utils/HibernateUtil.java [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+	public static SessionFactory getSessionFactory() {
+
+		return sessionFactory;
+	}
+
+	public static void shutdown(){
+
+		getSessionFactory().close();
+	}
+}
```
Comment (ces-hongvo) [inline src/net/bookmanagement/utils/HibernateUtil.java:39]: Update the order of the modifications

**Review by ces-hongvo — CHANGES_REQUESTED** [review]
@tototrinh 
I have left some comments. Please be noted
Thank you

**Comment by ces-totrinh** [comment]
@hongvo2308  I edited according to your comment. please check it again!! thank you

File: src/net/bookmanagement/servlet/BookServlet.java [hunk 14]
```diff
@@ -15,136 +15,144 @@
...
+			for (String n : authorNames) {
+				// get author object if it existed
+				Author author = authorDao.getAuthorByName(n);
+				if (author == null)
+					author = new Author(n);
+				authors.add(author);
+			}
+			// get category object if it existed
+			Category category = categoryDao.getCategoryByName(categoryName);
+			if (category == null)
```
Comment (ces-hongvo) [inline src/net/bookmanagement/servlet/BookServlet.java:91]: please use the brackets {}

File: src/net/bookmanagement/servlet/BookServlet.java [hunk 15]
```diff
@@ -15,136 +15,144 @@
...
+
+		for (String n : authorNames) {
+			Author author = authorDao.getAuthorByName(n);
+			if (author == null)
+				author = new Author(n);
+			authors.add(author);
+		}
+
+		Category category = categoryDao.getCategoryByName(categoryName);
+		if (category == null)
```
Comment (ces-hongvo) [inline src/net/bookmanagement/servlet/BookServlet.java]: Please use the brackets {}

**Review by ces-hongvo — CHANGES_REQUESTED** [review]
@tototrinh 
This PR is fine. Please be noted to use the brackets {} for some condition checking. That might help to read the code better.
Thank you

**Pull request merged**